chore: allow dataclasses >= 0.6, < 0.9

bastienboutonnet opened the pull request:
resolves #3150 

### Description
As asked in #3150 `dataclasses` could be allowed to vary more to make dbt cohabit with other projects which rely on version of dataclasses >0.6.

@jtcohen6 thanks for the pointers in the ticket

A long time ago when I contributed to dbt I remember the CI doesn't run on external PRs is that still the case? If so, how should we go about checking that this change doesn't break as you instructed in #3150 ?

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
    - well, i've been able to add dbt as a dependency on [dbt-sugar](https://github.com/bitpicky/dbt-sugar) via poetry successfully because the resolver isn't upset anymore ✨ 
 - [ ] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

gshank commented:
This pull request isn't mergeable because it points to your private versions of hologram. I guess you just want to see if the CircleCI tests pass?

bastienboutonnet commented:
@gshank yes that's what @jtcohen6 advised to do. So it will go like this:
1. get the hologram PR merged: https://github.com/fishtown-analytics/hologram/pull/42
2. if we're good here (which it looks it is) we can point to the newer version of hologram (I think I'll have to wait for a green light to know if a new version tag has been generated and repo is available on PyPI
3. check all works fine still and then you folks should be able to merge it.

bastienboutonnet commented:
@gshank ah I see you merged the hologram PR. Let me know when there's a new version that I can replace the pointer to fork with

gshank commented:
There's a new version, 0.0.14

bastienboutonnet commented:
Damn that's blazing fast @gshank I'll update in a wee bit.

bastienboutonnet commented:
@gshank alright change done.

Btw, I had a question in the PR description around this checklist item:
> I have updated the CHANGELOG.md and added information about my change to the "dbt next" section.
not sure if this is worthy of a changelog update but if so let me know and I can add a little thing

Let me know what you think and I can change accordingly.

bastienboutonnet commented on core/setup.py:
```diff
@@ -69,8 +69,8 @@ def read(fname):
         'werkzeug>=0.15,<2.0',
-        'dataclasses==0.6;python_version<"3.7"',
-        'hologram==0.0.13',
+        'dataclasses==0.7;python_version<"3.7"',
```
also actually I'm not sure if we should pin strictly to 0.7 here or let it be like the hologram depenency which is `>=0.6,<0.9`

jtcohen6 replied:
I agree with having this match the hologram dependency exactly: `'dataclasses>=0.6,<0.9;python_version<"3.7"'` feels right.

jtcohen6 reviewed (commented):
Nice quick work @bastienboutonnet! Thanks for the assist @gshank.

I definitely think this is worthy of a Changelog update + adding yourself to the list of Contributors :)

bastienboutonnet commented:
Thanks @jtcohen6 I added an entry in the changelog. Not sure if it will be going in `0.20.0` so I'm happy to move it. I couldn't find a `dbt next` section. Let me know if you think it fits better somewhere else.

jtcohen6 reviewed (approved):
Thanks for this @bastienboutonnet!
